HW2/P3/mandelbrot.pyx

```diff
@@ -26,30 +45,108 @@ cpdef mandelbrot(np.complex64_t [:, :] in_coords,
 
+    # get real and imaginary coords
+    c_reals = np.real(in_coords)
+    c_imags = np.imag(in_coords)
+
+
     with nogil:
-        for i in range(in_coords.shape[0]):
-            for j in range(in_coords.shape[1]):
-                c = in_coords[i, j]
-                z = 0
+        # multithread over rows i
+        for i in prange(in_coords.shape[0],
+                        schedule='static',
+                        chunksize=1,
+                        num_threads=4):
+
+            # assuming columns is multiple of 8
+            # take j as groups of 8. so j=0 is first group of 8,
+            # j=1 is second group of 8, etc
+            for j in range(in_coords.shape[1]/8):
+                # get group of 8 real and imaginary
+                c_real = AVX.make_float8(c_reals[i,8*j],
+                                         c_reals[i,8*j+1],
+                                         c_reals[i,8*j+2],
+                                         c_reals[i,8*j+3],
+                                         c_reals[i,8*j+4],
+                                         c_reals[i,8*j+5],
+                                         c_reals[i,8*j+6],
+                                         c_reals[i,8*j+7])
+
+                c_imag = AVX.make_float8(c_imags[i,8*j],
+                                         c_imags[i,8*j+1],
+                                         c_imags[i,8*j+2],
+                                         c_imags[i,8*j+3],
+                                         c_imags[i,8*j+4],
+                                         c_imags[i,8*j+5],
+                                         c_imags[i,8*j+6],
+                                         c_imags[i,8*j+7])
+
+                # z starts as 0's
+                z_real = AVX.make_float8(0,0,0,0,0,0,0,0)
+                z_imag = AVX.make_float8(0,0,0,0,0,0,0,0)
+                # tmp_counts starts as 0's
+                tmp_counts = AVX.make_float8(0,0,0,0,0,0,0,0)
+
+
+
+                # update z and counts
+                for iter in range(max_iterations):
+                    # update z = z*z + c
+                    # save copy of z_real as it is overwritten
+                    tmp = z_real
+
+                    z_real = AVX.add(c_real,
+                             AVX.sub(AVX.mul(z_real,z_real),
+                                     AVX.mul(z_imag,z_imag)))
+                    z_imag = AVX.add(c_imag,
+                             AVX.mul(AVX.float_to_float8(2.0),
+                                     AVX.mul(tmp,z_imag)))
+
+                    # make mask z < 4.0
+                    mask = AVX.less_than(AVX.add(AVX.mul(z_real,z_real),
+                                                 AVX.mul(z_imag,z_imag)),
+                                         AVX.float_to_float8(4.0))
+
+                    # if all z >= 4.0 stop the loop
+                    if AVX.signs(mask) == 0.0:
+                        break
```

Comment:
Using break greatly reduces the running time, even though we do not need this to get correct count.